### DISCUSSION TITLE
Save to mp4 directly?

### DISCUSSION THREAD

**Discussion by zesto01** [coldtype/coldtype#93] [General]
Hi, I saw how cool this project is and I was wondering if there was a way to directly save the animations directly to mp4 instead of the rendered images I saw in the tutorial. If there is a way, hopefully someone could redirect me the code. Thank you!

**Comment by stenson** [comment]
Thanks for the question! The easiest way to create an mp4 of a coldtype animation is to trigger the "release" function by hitting the `R` key while the viewer app is open. This will make an mp4 version of the animation and open the Finder (on mac) to show where the mp4 file was written.

A few caveats

- If you don't have ffmpeg installed, this functionality won't work (ffmpeg is not bundled with coldtype, so you'll need to install it on your system (on mac I use `brew` for that))
- You'll probably want to add `bg=1` and `render_bg=1` to your animation decorator (if they're not there already), since ffmpeg doesn't have great transparency compositing, as demonstrated in the first code snippet below

You can also customize the release action by writing your own release function in your coldtype source file, either by generating a partially-applied release function from an `@animation` object, or by writing a fully custom function.

Adding a partially-applied function to your source file would look like this:

```python
from coldtype import *

@animation(timeline=30, bg=hsl(0.65), render_bg=1)
def my_animation(f):
    return P(f.a.r.inset(100)).rotate(f.e(0, r=(0, 360)))

release = my_animation.export("h264", loops=1)
```

And a fully custom function would have this general form:

```python
def release(passes):
    print(passes) # can do something with this list of Frame references
```

There's an example of a custom release function here https://coldtype.goodhertz.com/tutorials/animation.html?highlight=release, and many of the examples here have custom release functions as well https://github.com/goodhertz/coldtype/tree/main/examples/animations, like [this one](https://github.com/goodhertz/coldtype/blob/main/examples/animations/recursive_text.py)

**Reply by zesto01** [reply]
Hi, thanks for the quick reply. I'm trying to run this on Google Colab and I think the viewer does not show and I will mainly rely on the release function to export the file, however when I run the partially applied function, I can't seem to find where it was saved.

**Reply by stenson** [reply]
ah gotcha, will submit another top-level answer so the colab answer doesn't get buried in this thread — basically render and release are one and the same by default on colab, since you have to "release" an animation (i.e. compile to mp4) to see it at all in the embedded colab environment

**Comment by stenson** [comment]
In the case of google colab, a call like `my_animation.render()` should do the trick, saving the file to the "local" file system in colab, as in the image below (i.e. hit the little folder icon in the left-hand sidebar to show the file storage). (You can also do `my_animation.render().show()` to get the file to display inline — if you do this, you can also download the rendered mp4 by clicking the `...` button on the embedded video and select download there.)

<img width="761" alt="image" src="https://user-images.githubusercontent.com/89770/158041446-6ceb6bc8-c010-4d3c-a7a3-bc8746935b5d.png">

(You can access the same notebook at this link: https://colab.research.google.com/drive/1sxNSdggg7mZmkQgSXG2WB2LwwtHA1UiK?usp=sharing)

**Reply by zesto01** [reply]
Alright, thanks for this!

**Comment by 16aces** [comment]
is "circle.render().show()"  called in the same file as the rest of the code or called from somewhere else?

**Reply by stenson** [reply]
`circle.render().show()` a specific to Google Colab, where there aren't really files per-se — are you asking about Colab in particular or the desktop coldtype viewer environment? If it's the latter and you have the viewer in the foreground, you can save an mp4 by hitting `a` to render all the frames, then hitting `r` to build an mp4 from those frames — let me know if that helps at all!

**Reply by 16aces** [reply]
I'm trying to automate something on a desktop. so hitting a and r isn't really an option. 
.render() works just fine in desktop mode but .show() throws a 
"[WinError 2] The system cannot find the file specified" error.   
I get the impression this wasn't developed with automation in mind, if that's the case don't worry about it it can use work with the PNGs from the .render() to build my own solution.
